design/1.0.md

```diff
@@ -0,0 +1,84 @@
+
+## Analysis Configuration
+
+Configuration will consist of the following key components:
+
+* Identification of sources, sinks, and sanitizers
+* Identification of analysis scope, e.g. to skip third-party or testing packages.
+* Specify diagnostics to be ignored, e.g false-positives
+
+### Identification of sources, sinks, and sanitizers
+
+Much of our work is directly with `ssa.Value`s and `ssa.Instruction`s.
+Whether those values or instructions represent one element or another,
+ the configuration to identify these members should be as uniform as possible.
+In that light, specification should follow:
+
+#### ValueSpecifier
+A ValueSpecifier marks an `ssa.Value`, e.g., for identification as a source or as the safe result of a sanitizer.
+Values should be specifiable via any or all of:
+
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+
+#### CallSpecifier
+A CallSpecifier marks an `*ssa.Call`, e.g., as a sink or sanitizer.
+Calls should be specifiable via any or all of:
+
+* Symbol: package path, function name, optional receiver name.
+* Context: A CallSpecifier indicating  
+* Based on arguments (ValueSpecifier)
+* Based on return values (ValueSpecifier)
+
+## Inference
+
+Inference represents the scope of what a user does *not* need to configure.
+For instance, we can currently detect getter methods and mark as propagating
+a marked field without requiring any explicit configuration.
+
+If a source type is aliased, the alias is properly identified as equivalent as the aliased source type.
+
+# Beyond a 1.0
+
+The following seem like significant improvements but seem larger in scope than what constitutes a stable 1.0.
+These items are open to interpretation and warrant discussion.
+
+## Improved Inference
+
+* Detect sanitization via zero-value.
+    * Explicit sanitization of `type Source struct{ data string }` is easily done via `s.data = ""`.
+    * Alternatively, a user may specify redaction placeholders, e.g. `"REDACTED"` rather than require a zero value.
+* A type defined from a source type (e.g., `type Foo core.Source`) should also be considered a source type.
+    * Note: as tags take part in type identify of the underlying struct literal, field tags are "inherited" in defined types based on existing types.
+      This may result in conflicting current behavior between those sources identified by tag and those identified by path/package/name.
+* A type which contains a known source type as a field should itself be a source type and associated field.  E.g., `type Wrapper struct { data core.Source }`
+* Current effort under the umbrella of *cross functional analysis* greatly refines inference of safety in scenarios that would currently produce a finding.
```

Comment:
Beyond cross function analysis, would we eventually want to get to *whole program* analysis? If so, we would need to discuss what that means in the context of our analysis. Here's an example scenario that I think we may eventually want to handle:
```
var s = core.Source{data: "SECRET"}

func main() {
  sanitizeSource() // this sanitizes the global
  sinkSource() // this sinks the global
}
```
This scenario is pretty terrible in terms of programming practices, but there might be situations where it would make sense for a user to have a global `Source`.

Reply:
> Beyond cross function analysis, would we eventually want to get to whole program analysis?

I meant this as a catch-all term.  Whole program analysis is a subset of cross-function analysis, right?

> If so, we would need to discuss what that means in the context of our analysis.

Unless we want to include it in 1.0, I think it would be premature to try to nail it down now.

Reply:
Hmm. The way I see it it's the other way around, haha! I agree with you that the distinction is not useful right now though.

Just for informational purposes, here is how I see it:
* Cross-function analysis = tracking values across function calls
* Whole-program analysis = tracking any and all values in an entire program, including e.g. globals